proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+  have GADTs but not type families in their programs; the blunderbuss
+  ``-XFancyTypes`` is too broad for them. Yet other languages have no
+  fine-grained mechanism for controlling language features. I think
+  this is a luxury we can live without, gaining more simplicity in
+  its place.
+
+* Because of the expressive power of language editions -- with their ability
+  to control all aspects of GHC's behavior -- we now have potentially an
+  even larger configuration space to worry about as implementers. ("Does
+  the bug happen with ``-XWeevils`` under ``Stable`` or just ``Experimental``?")
+  If we expected users to use language editions *and* the 100+ language
+  extensions, this would indeed be problematic. But, again, the goal is
+  to eschew language extensions for the coarser semantic bundles, reducing
+  the space down to something more manageable.
+
+* This proposal takes us further from where Haskell started, as a base
+  language supported by multiple compilers, each of which could introduce
+  its own extensions. While nothing stops another compiler from supporting
+  the same language editions as proposed here, the coarse nature of them
+  makes it harder to enter the field. Yet the reality is that GHC is the
+  only industrial strength Haskell compiler, and even it doesn't compile
+  standard Haskell (because of the ``Applicative`` superclass to ``Monad``,
+  at least).
+
+* If language editions come out only every three years, then there is more
+  pressure to get an idea into the next edition, instead of having to wait.
+  This is a real problem. It is mitigated a bit by the existence of ``Latest``,
+  which will evolve with every GHC release (possibly even minor releases). To
+  me, we're striking the right balance between due caution and our desire for
+  evolution, but individual experiences and opinions will differ.
+
+
+Backward Compatibility
+----------------------
+
+This proposal causes no change in behavior to existing programs, except
+for a warning that users should specify a language edition.
+
+Alternatives
+------------
+
+* The current proposal makes ``Stable2024`` and ``Experimental2024``
+  into language editions. In the future, we'll have, say, ``2027``
+  editions, too. A different way to slice this cake is to specify
+  the language varietal and its vintage separately, via two flags.
+
+  There's something appealing about the separation, as it makes
+  two axes navigable independently. Maybe it's a better user interface
+  than what I've proposed, but I think the payload doesn't change:
+  the combination of varietal and vintage would together make a language
+  edition, and all the aspects of a language edition that I've
+  described would be controlled by this little product.
+
+* ``Stable`` and ``Experimental`` are central to this proposal; ``Student``
+  and ``Latest`` less so. I can see forgoing either of these two. I like
+  keeping them, but maybe others disagree.
```

Comment:
I see very little reason to have `Latest`. We moved away from `-fglasgow-exts` a long time ago. Users can ask for the specific experimental/latest features they need, rather than getting an ad hoc collection of most of them.

I don't have a clear opinion about `Student`, although it feels rather arbitrary to me. I think it may be more confusing than helpful to have the available language features differ from `Stable`.